vision/diffusion_mnist/diffusion_plot.jl

```diff
@@ -0,0 +1,121 @@
+include("diffusion_mnist.jl")
+
+using Images
+
+"""
+Helper function yielding the diffusion coefficient from a SDE.
+"""
+function diffusion_coeff(t, sigma=25.0f0)
+    sigma .^ t
+end
+
+"""
+Helper function that produces images from a batch of images.
+"""
+function convert_to_image(x, y_size)
+    Gray.(
+        permutedims(
+            vcat(
+                reshape.(
+                    chunk(x |> cpu, y_size), 28, :
+                )...
+            ),
+            (2, 1)
+        )
+    )
+end
+
+"""
+Helper function that generates inputs to a sampler.
+"""
+function setup_sampler(model, device, num_images=5, num_steps=500, ϵ=1.0f-3)
+    t = ones(Float32, num_images) |> device
+    init_x = (
+        randn(Float32, (28, 28, 1, num_images)) .*
+        expand_dims(model.marginal_prob_std(t), 3)
+    ) |> device
+    time_steps = LinRange(1.0f0, ϵ, num_steps)
+    Δt = time_steps[1] - time_steps[2]
+    return time_steps, Δt, init_x
+end
+
+"""
+Sample from a diffusion model using the Euler-Maruyama method.
+
+# References
+https://yang-song.github.io/blog/2021/score/#how-to-solve-the-reverse-sde
+"""
+function Euler_Maruyama_sampler(model, init_x, time_steps, Δt, device)
+    x = mean_x = init_x
+    progress = Progress(length(time_steps))
+    @info "Start Euler-Maruyama Sampling"
+    for time_step in time_steps
+        batch_time_step = ones(Float32, size(init_x)[end]) .* time_step |> device
```

Comment:
batch_time_step = fill!(similar(init_x, size(init_x)[end]), 1) .* time_step

Reply:
Like this, because now we don't have to pass around `device`.